File: app/src/main/java/com/rayworks/droidweekly/EventAspect.java
package com.rayworks.droidweekly;

import com.rayworks.droidweekly.dashboard.TestEvent;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.MethodSignature;
import org.greenrobot.eventbus.EventBus;

import java.lang.reflect.Method;

import timber.log.Timber;

@Aspect
public class EventAspect {

    private static final String POINTCUT_METHOD =
            "execution(@org.greenrobot.eventbus.Subscribe * *(..))";

    @Pointcut(POINTCUT_METHOD)
    public void methodAnnotatedWithSubscribe() {
    }

    @Around("methodAnnotatedWithSubscribe()")
    public Object weaveJoinPoint(ProceedingJoinPoint joinPoint) throws Throwable {
        MethodSignature methodSignature = (MethodSignature) joinPoint.getSignature();
        String className = methodSignature.getDeclaringType().getSimpleName();
        String methodName = methodSignature.getName();

        Method method = methodSignature.getMethod();

        Object[] args = joinPoint.getArgs();
        if (args != null && args.length == 1) {
            if (args[0] instanceof TestEvent) {
                TestEvent event = (TestEvent) args[0];
                Class<?> clazz = event.getClass();

                Timber.i(">>> consuming test event ( name : %s), removed now",
                        clazz.getSimpleName());
                EventBus.getDefault().removeStickyEvent(clazz);
                EventBus.getDefault().removeStickyEvent(event);
            }
        }

        return joinPoint.proceed();
    }
}

File: app/src/main/java/com/rayworks/droidweekly/EventAspect.kt
package com.rayworks.droidweekly

import com.rayworks.droidweekly.dashboard.TestEvent
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.aspectj.lang.reflect.MethodSignature
import org.greenrobot.eventbus.EventBus
import timber.log.Timber
import java.util.*

@Aspect
class EventAspect {
    companion object {
        const val POINTCUT_METHOD = "execution(@org.greenrobot.eventbus.Subscribe * *(..))"
    }


    @Pointcut(POINTCUT_METHOD)
    fun methodAnnotatedWithSubscribe() {
    }

    @Around("methodAnnotatedWithSubscribe()")
    @Throws(Throwable::class)
    fun weaveJoinPoint(joinPoint: ProceedingJoinPoint): Any? {
        val methodSignature = joinPoint.signature as MethodSignature
        val className = methodSignature.declaringType.simpleName
        val methodName = methodSignature.name
        val method = methodSignature.method
        val args = joinPoint.args
        println(">>> ${Arrays.toString(args)}")

        if (args != null && args.size == 1) {
            if (args[0] is TestEvent) {
                val event: TestEvent = args[0] as TestEvent
                val clazz: Class<*> = event.javaClass
                Timber.i(">>> consuming sticky event ( name : %s ) in method : %s, removed now", clazz.simpleName, methodName)
                EventBus.getDefault().removeStickyEvent(clazz)
                EventBus.getDefault().removeStickyEvent(event)
            }
        }
        return joinPoint.proceed()
    }
}